### ISSUE TITLE
Job keys aren't deleted after execution when async=False

### ISSUE THREAD

**Issue by ryanisnan** [nvie/rq#516]
I'm using RQ 0.5.2 and I've noticed that keys are not expiring and the size of my redis store is growing and growing. I have another project running 0.4.6 and I've never experienced this issue before.

Any thoughts?

**Comment by selwin** [comment]
We inteoduced a job.ttl feature in the later versions of RQ so there maybe bugs that cause job keys to never expire. Do you mind making a failing test case? Or alternatively, what parameters are you using to enqueue your jobs?

**Comment by ryanisnan** [comment]
We're using https://github.com/ui/django-rq and its `@job` decorator to define jobs. From what I can tell (https://github.com/ui/django-rq/blob/master/django_rq/decorators.py) it looks like the parameters are all default. We're not passing anything aside from the queue name.

**Comment by selwin** [comment]
Hmm, we have a test that ensures correct ttl values are set after jobs are successfully executed here: https://github.com/nvie/rq/blob/master/tests/test_worker.py#L201

Are your jobs failing by any chance? Failed jobs are not deleted.

**Comment by ryanisnan** [comment]
There is a very small number of jobs that are failing, but the vast majority of them are succeeding (18 failures over the last 24 hours).

**Comment by selwin** [comment]
Could these be the ones that cause your key count to rise? You can monitor and delete them from django-rq's administration panel.

**Comment by ryanisnan** [comment]
So I'm going to pre-emptively ask you for your forgiveness. It turns out that I made a mistake, and somehow shipped an ASYNC flag up to production. It appears that when ASYNC mode is enabled, the items are added somewhere, but never removed?

**Comment by selwin** [comment]
Ah ok, this explains why no one has complained about this, even though this is a bug. I'd imagine that not many people run RQ in synchronous mode in production ;).

It's still a bug though so I'll change the title to reflect what the issue really is. Thanks for the report.

**Comment by ryanisnan** [comment]
I'll take a look at RQ source to see if I can lend a hand in fixing it!

**Comment by hugorodgerbrown** [comment]
Hi there - we've just run into exactly the same issue (and same cause - inadvertently running one type of job with `async=False`). I've raised an issue over on django_rq - https://github.com/ui/django-rq/issues/115 - which I will close in favour of this one.

**Comment by hugorodgerbrown** [comment]
tl;dr on the other issue - it's because whilst the `worker.perform_job` method [sets the state of the job to 'finished'](https://github.com/nvie/rq/blob/master/rq/worker.py#L567-L585), the code inside `queue.enqueue_job` [that runs when in sync mode](https://github.com/nvie/rq/blob/master/rq/queue.py#L276-L278) does not, and so the jobs never get updated / cleared out.

**Comment by selwin** [comment]
Fixed in https://github.com/nvie/rq/pull/561

**Comment by hugorodgerbrown** [comment]
Fantastic - thanks.
